### PR TITLE
Push notifications API implementation

### PR DESCRIPTION
**Pull request by RishiDiwanTT**
## Description
- model for DeviceTokens
  which stores the FCM token for patrons mobile devices
  It ensure uniqueness on the token string

- API to GET and PUT device tokens for patrons
- GET
  responds with 200 and token details
  404 if token is not found
- PUT
  responds with 201 if created
  400 if the token type is invalid
  409 if the token was not unique
- There is no DELETE method yet
  No use case is outlined for it

## Motivation and Context
The Palace Manager needs to support the following push notifications in the iOS and Android apps
This is the first step, allowing the apps to store the FCM tokens for the mobile devices

## How Has This Been Tested?
Unit tests were written and manual testing was done via Postman

## Checklist

- [ ] I have updated the documentation accordingly.
- [x] All new and existing tests passed.

### CODE REVIEW

File: core/model/devicetokens.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+from . import Base
+
+
+class DeviceTokenTypes:
+    FCM_ANDROID = "FCMAndroid"
+    FCM_IOS = "FCMiOS"
+
+
+T = TypeVar("T", bound="DeviceToken")
```
Comment (io7m) [inline core/model/devicetokens.py:17]: :+1: :+1: :+1: :+1:

**Review by io7m — APPROVED** [review]
This looks good.

**Comment by io7m** [comment]
I like the use of generics. Most people don't think of them!

File: core/model/devicetokens.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+    device_token = Column(Unicode, nullable=False, unique=True, index=True)
+
+    @classmethod
+    def create(
+        cls: Type[T],
+        db,
+        token_type: str,
+        device_token: str,
+        patron: "Union[Patron,int]",
```
Comment (tdilauro) [inline core/model/devicetokens.py:43]: Do we need to quote the type here?

Comment (RishiDiwanTT) [reply core/model/devicetokens.py:43]: Patron was earlier imported under `if Typing` I can change it to unquoted if needed

File: core/model/devicetokens.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        elif type(patron) is Patron:
+            kwargs["patron_id"] = patron.id
+
+        device = cls(**kwargs)
+        try:
+            db.add(device)
+            db.commit()
+        except IntegrityError as e:
+            if "device_token" in e.args[0]:
+                raise DuplicateDeviceTokenError()
```
Comment (tdilauro) [inline core/model/devicetokens.py:63]: Use raise from here: `raise DuplicateDeviceTokenError() from e`?

Comment (RishiDiwanTT) [reply core/model/devicetokens.py]: Done

**Pull request merged**